### PR TITLE
Deduplicate marker symbols

### LINKED ISSUE #222: Redundant Symbols in markers layer
## 🐛 Bug

Some symbols in the markers layer are just the sam, or at least I can't make the difference. For example: '|' and 'vbar' appear to be the same, other instances: 'o' and 'disc'.

image

## To Reproduce

N/A

## Expected behavior

Cleaner list of symbols without duplicates.

## Environment

 - napari Version (e.g., 1.0): latest dimslider PR
 - OS (e.g., Linux): OSX
 - Python version: 3.7
 - Any other relevant information:

## Additional context

N/A

### PR DESCRIPTION
**Pull request by kevinyamauchi**
# Description
As raised by @royerloic in #222 , the markers layer properties panel has redundant options for the marker symbol types. This PR deduplicates the list and adds an Enum (`Symbols`) for storing the deduplicated list of symbol names and returning the valid vispy symbol names. Users can programmatically set the marker symbol either via the vispy string or the `Symbols` Enum.

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [x] This change requires a documentation update

# References
This PR closes #222.

# How has this been tested?
- [x] I ran the add_markers.py example.

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: napari/layers/_markers_layer/_constants.py
Comment (kne42) [inline napari/layers/_markers_layer/_constants.py]: IMO you should be using strings instead of integers but it's not a big deal

File: napari/layers/_markers_layer/model.py
Comment (kne42) [inline napari/layers/_markers_layer/model.py]: IMO a better pattern is:
```python
if isinstance(symbol, str):
    symbol = Symbols[symbol.upper()]
self._symbol = symbol
```

File: napari/layers/_markers_layer/view/properties.py
Comment (kne42) [inline napari/layers/_markers_layer/view/properties.py]: is this necessary since `upper` will be called on it anyways when it gets converted to an Enum?

Comment (kevinyamauchi) [reply napari/layers/_markers_layer/view/properties.py]: Ah good catch!

File: napari/layers/_markers_layer/_constants.py
Comment (kevinyamauchi) [reply napari/layers/_markers_layer/_constants.py]: I went back and forth on strings vs. integers. Ultimately, I settled on integers because it makes it easy to get the index for updating the symbol type drop down. If you  or anybody has a strong preference, I am happy to change it.

Comment (kne42) [reply napari/layers/_markers_layer/_constants.py]: Yeah, I saw that later in the code. Definitely makes sense 👍

File: napari/layers/_markers_layer/model.py
Comment (kne42) [inline napari/layers/_markers_layer/model.py]:
```suggestion
    symbol : Symbols or {'arrow', 'clobber', 'cross', 'diamond', 'disc',
                         'hbar', 'ring', 'square', 'star', 'tailed_arrow', 
                         'triangle_down', 'triangle_up', 'vbar', 'x'}
```
and remove that information from the description

File: napari/layers/_markers_layer/_constants.py
Comment (kne42) [inline napari/layers/_markers_layer/_constants.py]: I think `Symbol` makes more sense. @jni what are your thoughts on plural vs singular enum family names?

File: napari/layers/_markers_layer/model.py
Comment (kne42) [inline napari/layers/_markers_layer/model.py]: shouldn't this be lowercase? and what was wrong with having `o` as a valid option, I don't think we should change that

Comment (kevinyamauchi) [reply napari/layers/_markers_layer/model.py]: The case doesn't really matter because I make them all upper case in the `Markers.symbol` setter, but it definitely looks better lower case so I will change it.

In terms of having `o` as a valid option, would you be okay with removing disc in favor of keeping `o` (`o` and `disc` map to the same symbol)?

File: napari/layers/_markers_layer/_constants.py
Comment (kevinyamauchi) [reply napari/layers/_markers_layer/_constants.py]: Good point. I looked around a bit and singular definitely looks more standard. I went ahead and changed the name to singular. @jni, let me know if you have a strong preference the other way.

File: napari/layers/_markers_layer/model.py
Comment (kne42) [inline napari/layers/_markers_layer/model.py]: If you are going forward with enums, this should at least still return a string.
```suggestion
         return str(self._symbol)
```

**Review by kne42 — CHANGES_REQUESTED** [review]
Actually looking back at the reasoning behind this I fundamentally disagree with how it is implemented because it worsens the API. Also enums are the worst and I don't understand why anyone would want more of them unless it's absolutely necessary and justified but I guess it's ok to keep them around if people _really_ want it.

I think that the idea behind the enum was good in that you deduplicated the symbols' aliases and that way it makes the drop-down less confusing. What I don't like is that all of these aliases were lost from the programmatic interface. So I would add some sort of conversion from the aliases to the symbols.

File: napari/layers/_markers_layer/_constants.py
Comment (kne42) [inline napari/layers/_markers_layer/_constants.py]: There's no new information in this link so it should be removed.

File: napari/layers/_markers_layer/model.py
Comment (kevinyamauchi) [reply napari/layers/_markers_layer/model.py]: Thanks for the feedback. If I understand your comment, your primary concerns are: (1) you want to keep the ability to set marker symbols with the alternative names via `Markers.symbol` and (2) you don't think using an Enum to hold the deduplicated names is useful.

- Would you prefer that we hold the deduplicated names for the UI as strings in a list or tuple in the _constants.py file (edit: or as an attribute of `Markers` per your comment in #222 )? This is a straightforward change, we just have to string match the updated `Markers.symbol` value and the names in the UI dropdown (this is how it used to be).
- We can include a dictionary to map the alias to our deduplicated symbol names in the setter. Do you think this would be okay?

**Comment by royerloic** [comment]
> Actually looking back at the reasoning behind this I fundamentally disagree with how it is implemented because it worsens the API. Also enums are the worst and I don't understand why anyone would want more of them unless it's absolutely necessary and justified but I guess it's ok to keep them around if people _really_ want it.
> 
> I think that the idea behind the enum was good in that you deduplicated the symbols' aliases and that way it makes the drop-down less confusing. What I don't like is that all of these aliases were lost from the programmatic interface. So I would add some sort of conversion from the aliases to the symbols.

Ok, I am going to strongly disagree here. Enums are great. Strings have barely semantics above comments. You can happily pass a string that is unrelated and there is no type checking. Strings are great for a script based coding, but terrible if you actually want to build solid stateful software. As Napari grows rapidly, enums provide an explicit typing and document implicitly what these parameter values can be.  This is very important for long term stability of the code.

I have nothing against an external facing API that uses strings -- and Python has enough metaprogramming features to pull this. But, internally, we need something that machines can reason with (enums) instead of something that only makes sense to humans (strings). This is also usefull when using refactoring tools in IDE such as Pycharm etc...

We should definitely be using enums as much as possible internally and figure out how to provide a convenient conversion with the external API. But internally, I should be able to figure out what are valid values for a parameter and not have to hunt for some strings here and there...

**Comment by kne42** [comment]
> You can happily pass a string that is unrelated and there is no type checking.

Type checking is not really a concept in Python so I don't get your point here. It's no different than if you passed a tuple in instead and the function errored. Garbage in, garbage out.

> Strings are great for a script based coding, but terrible if you actually want to build solid stateful software. As Napari grows rapidly, enums provide an explicit typing and document implicitly what these parameter values can be. This is very important for long term stability of the code.

These parameters' values are already documented inside the function docstring and additionally can be looked up since they are still stored in some sort of collection as valid values. Excessive type-checking is fundamentally unpythonic and not necessary for stability since proper following of function signatures is just as effective.

> But, internally, we need something that machines can reason with (enums) instead of something that only makes sense to humans (strings).

I don't understand what you mean here. Machines can reason equally well with strings and enums.

> This is also usefull when using refactoring tools in IDE such as Pycharm etc...

Is it worth making the codebase unnecessarily messy/complicated just so that you can refactor a little bit easier?

> I should be able to figure out what are valid values for a parameter and not have to hunt for some strings here and there...

There won't be a hunt if the information is all in the documentation. Sure, maybe your life will be a tad bit harder because you can't just use PyCharm to autoimport and autocomplete your enums but I don't think it is good practice for public function to accept enums in the first place.

File: napari/layers/_markers_layer/model.py
Comment (kne42) [reply napari/layers/_markers_layer/model.py]: 1. yeah preferably stored as strings in a tuple as a class variable
2. yeah that sounds great

**Comment by royerloic** [comment]
Type checking is fundamental to proper software engineering,
and it is interesting to see Python moving towards having more 
types... because it is useful... Not always, not in every situation,
but for large complex projects it is essential.  Pycharm can check
some amount of types at edit time, and python support for type hints
is getting better.  

About refactoring: this is the most important concept 
in software engineering. Ask anybody that has a few 
decades of experience on large projects. I would
take refactoring anytime. The main reason is that you
don't type much anymore if you have good tools anyway.
And you can handle incredibly complex projects seamlessly.
As soon as you handle/refactor projects with several hundred files 
you will realise how essential it is and how important to have 
code that accommodates tooling. 
What I meant with code understandable by machines is code
that can be analysed and processed with tools like refactoring 
IDEs and such...  

I rest my case :-)

**Review by royerloic — DISMISSED** [review]
Enums...

**Comment by kevinyamauchi** [comment]
I think the public-facing API and the internal representation of the symbol marker names can largely be two parallel discussion.

I agree with @kne42 that as a user, being required to use custom Enums to set parameters is really annoying. The current version of the branch accepts either strings of the symbol names or `Symbol` to set the `Markers.symbol`. As discussed above, we can further extend it to include strings of all marker symbol names accepted by vispy.

Regarding using an Enum to store the symbol marker names, I was thinking along the same lines as @royerloic . However, I am new to Python, so if they don't have utility here I am happy to switch to a tuple or list.

**Comment by royerloic** [comment]
Strings for the front-facing user API, enums internally, that would be the best.

**Comment by kne42** [comment]
Sure, enums are great for refactoring. In massive projects. Where they are used all over the place. _That_'s . proper justification to me.

You still haven't provided any justification to me for use of an enum just for something like this that's likely going to be pretty small in scope.

File: napari/layers/_markers_layer/view/properties.py
Comment (kne42) [inline napari/layers/_markers_layer/view/properties.py:66]: since you have found a way to get the index without needing to use the enum's value, might you consider setting the enum values to the string forms of their names?

File: napari/layers/_markers_layer/_constants.py
Comment (kne42) [inline napari/layers/_markers_layer/_constants.py]: nice! since we have agreed to use enums going forward, these aliases should translate back to enums

also you should have two lines of whitespace between your import statements and your first line of code

**Review by kne42 — APPROVED** [review]
approved pending resolution of below comments

File: napari/layers/_markers_layer/_constants.py
Comment (kevinyamauchi) [reply napari/layers/_markers_layer/_constants.py]: Good call. Done.

File: napari/layers/_markers_layer/view/properties.py
Comment (kevinyamauchi) [reply napari/layers/_markers_layer/view/properties.py:66]: That makes sense to me. Will do.

**Comment by kevinyamauchi** [comment]
@royerloic, as we discussed, we are using an Enum for the deduplicated marker symbols and are using strings for the public facing interface. Kira has approved. Is there anything else you would like me to change before merging?

**Review by sofroniewn — APPROVED** [review]
This looks good to me. I think we should use this system of `str` with `Enum` wherever we have `Enum` in the code base right now, i.e. the layer Modes, and we should also curate our own list of `colors` and `colormaps` with `Enums`, but these can be follow up PRs

**Comment by kevinyamauchi** [comment]
I totally agree, @sofroniewn. I'll try to compile a list of places we need to unify the Enum behavior/usage and then post an issue.

@royerloic, since Nick and Kira approve and we went with Enums as you asked, I will go ahead and merge unless I hear back from you by tonight.

File: napari/layers/_markers_layer/_constants.py
Comment (kne42) [inline napari/layers/_markers_layer/_constants.py]:
```suggestion
        return self.value
```

File: napari/layers/_markers_layer/_constants.py
Comment (kne42) [inline napari/layers/_markers_layer/_constants.py]: this (and the ones below) should look like `Symbol.DISC`

File: napari/layers/_markers_layer/model.py
Comment (kne42) [inline napari/layers/_markers_layer/model.py]:
```suggestion
    def __init__(self, coords, symbol='o', size=10, edge_width=1,
```

**Review by kne42 — CHANGES_REQUESTED** [review]

File: napari/layers/_markers_layer/model.py
Comment (kne42) [inline napari/layers/_markers_layer/model.py]:
```suggestion
                symbol = Symbol(symbol)
```

File: napari/layers/_markers_layer/_constants.py
Comment (kevinyamauchi) [reply napari/layers/_markers_layer/_constants.py]: Done!

**Review by kne42 — APPROVED** [review]

**Pull request merged**
